### ISSUE TITLE
Atom object returned NULL without setting an error

### ISSUE THREAD

**Issue by blakejohnson** [nucleic/enaml#216]
As first reported here https://github.com/nucleic/enaml/issues/117#issuecomment-261643802.

I've created a (somewhat) reduced test case:
https://gist.github.com/blakejohnson/b60fea03e6db96d359d39e1757e6c67b

You run the test with
```shell
shell$ python MeasFilters.py
```
Then select any item from the list on the left side of the window and click the "remove" button.

On a python 2.7 (Anaconda 4.2.0) install with `conda install enaml`, this works fine.

On python 3.5 (Anaconda 4.2.0) with `conda install -c ecpy enaml`, you get an error like:
```
Traceback (most recent call last):
  File "/Users/bjohnson/anaconda3/envs/enaml/lib/python3.5/site-packages/enaml/qt/qt_abstract_button.py", line 76, in on_clicked
    self.declaration.clicked(checked)
  File "/Users/bjohnson/anaconda3/envs/enaml/lib/python3.5/site-packages/enaml/core/declarative_meta.py", line 68, in declarative_change_handler
    engine.write(owner, change['name'], change)
  File "/Users/bjohnson/anaconda3/envs/enaml/lib/python3.5/site-packages/enaml/core/expression_engine.py", line 210, in write
    pair.writer(owner, name, change)
  File "/Users/bjohnson/anaconda3/envs/enaml/lib/python3.5/site-packages/enaml/core/standard_handlers.py", line 82, in __call__
    call_func(func, (), {}, scope)
  File "/Users/bjohnson/src/enaml-issue/DictManagerView.enaml", line 47, in <module>
    clicked :: dictManager.remove_item(itemList.selected_item)
  File "/Users/bjohnson/src/enaml-issue/DictManager.py", line 42, in remove_item
    self.displayList.remove(itemLabel)
SystemError: <built-in method remove of atom.catom.atomclist object at 0x10550e990> returned NULL without setting an error
```

The problem seems to be that `itemLabel` is no longer in `self.displayList` by the time the `remove` method is fired.  This would be true if the `itemDict` observer fired prior to executing this command. However, at least in the python 2.7 version of atom/enaml, that event only fires when the entire `itemDict` variable is rebound, and not on a `pop`.

**Comment by MatthieuDartiailh** [comment]
Thanks for sharing this !

**Comment by MatthieuDartiailh** [comment]
Actually you forgot a xrange in qt_list_str_widget which hence raise (using range allow everything to work fine). The issue is that this should lead to a NameError not a SystemError. I will try to use this example to investigate more.

**Comment by blakejohnson** [comment]
Thanks for finding that bug, @MatthieuDartiailh!

**Comment by MatthieuDartiailh** [comment]
I finally managed to trim down a bit the reproduction example. Follow the gist exhibiting the issue.
https://gist.github.com/MatthieuDartiailh/e44b22316b582f3242f27b910a74be96

I created some custom branches of atom and enaml in my forks (debug-enaml in both cases) adding ugly print for debugging. I reached a weird point where the error seem to disappear between to c++ calls which makes no sense (ObserverPool.notify test before returning PyErr_Occurred and according to the message find one exception, reading the code it should have been invoked from CAtom.notify in which the same PyErr_Occurred show no exception). Below is the console output. @sccolbert do you have any idea of what may be happening here ? 
```
enaml-run test_bug.enaml 
QGtkStyle could not resolve GTK. Make sure you have installed the proper libraries.
Standard traced handler <__main__.QtListStrWidget object at 0x7fcf1c826f78> items
Standard traced handler return <__main__.QtListStrWidget object at 0x7fcf1c826f78> items []
SubscriptionObserver <__main__.QtListStrWidget object at 0x7fcf1c826f78> items
Enter ExpressionEngine.update for <__main__.QtListStrWidget object at 0x7fcf1c826f78>, items
PairReader <enaml.core.standard_handlers.StandardTracedReadHandler object at 0x7fcf1c816748>
Standard traced handler <__main__.QtListStrWidget object at 0x7fcf1c826f78> items
Standard traced handler return <__main__.QtListStrWidget object at 0x7fcf1c826f78> items ['e']
PairReader returned
Fail at notify (because of static)
Exception in ExpressionEngine.update for  <__main__.QtListStrWidget object at 0x7fcf1c826f78> items RuntimeError()
Exception in SubscriptionObserver <__main__.QtListStrWidget object at 0x7fcf1c826f78> items RuntimeError()
Fail at ObserverPool.notify: Python exception set
CAtom: Fail at dynamic notifier: Python exception not set
MemberSetattrBehavior: Fail at notify (because of dynamic): Python exception not set
Traceback (most recent call last):
  File "/home/marul/Documents/coding/enaml/enaml/qt/qt_abstract_button.py", line 76, in on_clicked
    self.declaration.clicked(checked)
  File "/home/marul/Documents/coding/enaml/enaml/core/declarative_meta.py", line 68, in declarative_change_handler
    engine.write(owner, change['name'], change)
  File "/home/marul/Documents/coding/enaml/enaml/core/expression_engine.py", line 210, in write
    pair.writer(owner, name, change)
  File "/home/marul/Documents/coding/enaml/enaml/core/standard_handlers.py", line 82, in __call__
    call_func(func, (), {}, scope)
  File "test_bug.enaml", line 94, in <module>
    main.items = l
SystemError: error return without exception set
```

**Comment by MatthieuDartiailh** [comment]
Actually using my debug branches show that this bug may be related https://github.com/MatthieuDartiailh/enaml/issues/5. (traceback below showing the same disappearance of the expected ValueError in the exact same place). If it is so solving the above bug will fix all the known bugs of enaml on Python 3 !

```
QGtkStyle could not resolve GTK. Make sure you have installed the proper libraries.
Standard traced handler <enaml.widgets.field.Field object at 0x7f83c8e35948> text
Standard traced handler return <enaml.widgets.field.Field object at 0x7f83c8e35948> text 
Standard traced handler <enaml.widgets.field.Field object at 0x7f83c8e35988> text
Standard traced handler <__main__.Main object at 0x7f83c8e358c8> bis
Standard traced handler return <__main__.Main object at 0x7f83c8e358c8> bis 
Standard traced handler return <enaml.widgets.field.Field object at 0x7f83c8e35988> text 
Standard traced handler <enaml.widgets.field.Field object at 0x7f83c8e359c8> text
Standard traced handler <__main__.Main object at 0x7f83c8e358c8> ter
Standard traced handler return <__main__.Main object at 0x7f83c8e358c8> ter 
Standard traced handler return <enaml.widgets.field.Field object at 0x7f83c8e359c8> text 
SubscriptionObserver <enaml.widgets.field.Field object at 0x7f83c8e35948> text
Enter ExpressionEngine.update for <enaml.widgets.field.Field object at 0x7f83c8e35948>, text
SubscriptionObserver <__main__.Main object at 0x7f83c8e358c8> bis
Enter ExpressionEngine.update for <__main__.Main object at 0x7f83c8e358c8>, bis
PairReader <enaml.core.standard_handlers.StandardTracedReadHandler object at 0x7f83cde7d148>
Standard traced handler <__main__.Main object at 0x7f83c8e358c8> bis
Standard traced handler return <__main__.Main object at 0x7f83c8e358c8> bis d
PairReader returned
Bis event handler d
SubscriptionObserver <enaml.widgets.field.Field object at 0x7f83c8e35988> text
Enter ExpressionEngine.update for <enaml.widgets.field.Field object at 0x7f83c8e35988>, text
PairReader <enaml.core.standard_handlers.StandardTracedReadHandler object at 0x7f83cde81fc8>
Standard traced handler <enaml.widgets.field.Field object at 0x7f83c8e35988> text
Standard traced handler return <enaml.widgets.field.Field object at 0x7f83c8e35988> text d
PairReader returned
SubscriptionObserver <__main__.Main object at 0x7f83c8e358c8> ter
Enter ExpressionEngine.update for <__main__.Main object at 0x7f83c8e358c8>, ter
PairReader <enaml.core.standard_handlers.StandardTracedReadHandler object at 0x7f83cde815c8>
Standard traced handler <__main__.Main object at 0x7f83c8e358c8> ter
Standard traced handler return <__main__.Main object at 0x7f83c8e358c8> ter dd
PairReader returned
Ter event handler dd
Fail at notify (because of static)
Exception in ExpressionEngine.update for  <__main__.Main object at 0x7f83c8e358c8> ter ValueError()
Exception in SubscriptionObserver <__main__.Main object at 0x7f83c8e358c8> ter ValueError()
Fail at ObserverPool.notify: Python exception set
CAtom: Fail at dynamic notifier: Python exception not set
MemberSetattrBehavior: Fail at notify (because of dynamic): Python exception not set
Fail at notify (because of static)
Traceback (most recent call last):
  File "/home/marul/Documents/coding/enaml/enaml/qt/qt_field.py", line 147, in on_submit_text
    self._validate_and_apply()
  File "/home/marul/Documents/coding/enaml/enaml/qt/qt_field.py", line 101, in _validate_and_apply
    d.text = text
  File "/home/marul/Documents/coding/enaml/enaml/core/declarative_meta.py", line 68, in declarative_change_handler
    engine.write(owner, change['name'], change)
  File "/home/marul/Documents/coding/enaml/enaml/core/expression_engine.py", line 210, in write
    pair.writer(owner, name, change)
  File "/home/marul/Documents/coding/enaml/enaml/core/standard_handlers.py", line 123, in __call__
    call_func(func, (inverter, change['value']), {}, scope)
  File "test_scope_bug.enaml", line 21, in <module>
    text := aux
  File "/home/marul/Documents/coding/enaml/enaml/core/standard_inverter.py", line 38, in load_name
    self.scope['nonlocals'][name] = value
KeyError: 'aux'
```

**Comment by MatthieuDartiailh** [comment]
Going deeper I discovered that : 
- bypassing the pair.reader in ExpressionEngine.update allow to recover the proper error
- looking at what happens inside the reader shows nothing more than basic tracing
- looking at the tracer finalisation observers can be invalidated or added, and replacing the invalidation by a full removal allow to get back the proper error

So I guess that the issue comes from the cleanup of the GuardMap or something similar. @sccolbert does that make sense to you ?

I update my debug branches to show this behavior.

**Comment by MatthieuDartiailh** [comment]
I finally pin-pointed the root of the issue. It turns out that under Python 3 if an exception is set before testing the truth value of a SubscriptionObserver (while destroying the GuardMap and hence adding an observer) using PyObject_IsTrue returns -1. The PyObjectPtr.is_true method subsenquently clears the exception.
I believe that the proper fix for the issue would to fetch the error at the beginngng of the GuardMap destroyer and restore it before returning. My reasoning here is that the destroyer should behave like a finally clause.
Another option would be to add the logic in is_true but I dislike this option.
@sccolbert I will try to implement it and update the Atom PR subsequently.

**Comment by blakejohnson** [comment]
nice detective work

**Comment by sccolbert** [comment]
Nice work indeed. Lemme ponder on the correct fix for this. I really appreciate all the effort you are putting in, and I'm terribly sorry it's taking me so long to address these issues.

**Comment by MatthieuDartiailh** [comment]
Fixed in atom 0.4.0
